src/services/YotiService.ts

```diff
@@ -31,309 +31,330 @@ export class YotiService {
+    		}
+    	} else {
+    		this.logger.error({ message: "Missing Yoti request config ", yotiRequest });
+    	}
 	}
 
-	async getCompletedSessionInfo(sessionId: string): Promise<YotiCompletedSession | undefined> {
-		const yotiRequest = this.generateYotiRequest({
-			method: HttpVerbsEnum.GET,
-			endpoint: `/sessions/${sessionId}`,
-		});
-
-		try {
-			const { data } = await axios.get(yotiRequest.url, yotiRequest.config);
-
-			return data;
-		} catch (error: any) {
-			const xRequestId = error.response ? error.response.headers["x-request-id"] : undefined;
-			this.logger.error({ message: "An error occurred when fetching Yoti session", yotiErrorMessage: error.message, yotiErrorCode: error.code, messageCode: MessageCodes.FAILED_YOTI_GET_SESSION, xRequestId });
-			throw new AppError(HttpCodesEnum.SERVER_ERROR, "Error fetching Yoti Session");
-		}
+	async getCompletedSessionInfo(sessionId: string, backoffPeriodMs: number, maxRetries: number): Promise<YotiCompletedSession | undefined> {
+    	const yotiRequest = this.generateYotiRequest({
+    		method: HttpVerbsEnum.GET,
+    		endpoint: `/sessions/${sessionId}`,
+    	});
+    	let retryCount = 0;
+    	while (retryCount <= maxRetries) {
+    		this.logger.debug({
+    			message: "getCompletedSessionInfo - trying to fetch Yoti session", 
+    			yotiSessionId: sessionId,
+    			retryCount,
+    		});
+			console.log("retryCount: " + retryCount + ", backoffPeriodMs: " + backoffPeriodMs + "retries: " + maxRetries);
+
+    		try {
+    			const { data } = await axios.get(yotiRequest.url, yotiRequest.config);
+
+    			return data;
+    		} catch (error: any) {
+    			const xRequestId = error.response ? error.response.headers["x-request-id"] : undefined;
+				
+    			if (((error.response?.status >= 500 && error.response?.status < 600) || error.response?.status === 429) && retryCount < maxRetries) {
+    				this.logger.warn({ message: `getCompletedSessionInfo - Retrying to fetch Yoti session. Sleeping for ${backoffPeriodMs} ms`, retryCount, yotiErrorMessage: error.message, yotiErrorCode: error.code, yotiErrorStatus: error.response?.status, messageCode: MessageCodes.FAILED_YOTI_GET_SESSION, xRequestId });
+    				await sleep(backoffPeriodMs);
+    				retryCount++;
+    			} else {
+					if (retryCount === maxRetries) {
+						this.logger.error({ message: `getCompletedSessionInfo - cannot fetch Yoti session even after ${maxRetries} retries.`, messageCode: MessageCodes.YOTI_RETRIES_EXCEEDED, xRequestId });
+    					throw new AppError(HttpCodesEnum.SERVER_ERROR, `Cannot fetch Yoti session even after ${maxRetries} retries.`);
+					}
+					const message = "An error occurred when fetching Yoti session";
+					this.logger.error({ message, yotiErrorMessage: error.message, yotiErrorCode: error.code, messageCode: MessageCodes.FAILED_YOTI_GET_SESSION, xRequestId });
+    				throw new AppError(HttpCodesEnum.SERVER_ERROR, message);
+    			}
+    		}
```

Comment:
This function is quite long, what do you think about extracting the error handling into it's own function that could maybe be reused in the future? 👀

Reply:
It will be difficult as we increment the retryCount var which is referenced in the while and will be too complex.

Reply:
Ok fair enough 👍